dynamiqs/sparse.py

```diff
@@ -0,0 +1,149 @@
+                )
+        return out
+    
+    @functools.partial(jax.jit, static_argnums=(0,1))
+    def _diamul(self, matrix) -> Array:
+        N = matrix.diags.shape[1]
+        
+        vector_list = []
+        offset_list = []
+
+        for offset, diag in zip(self.offsets, self.diags):
+            for matrix_offset, matrix_diag in zip(matrix.offsets, matrix.diags):
+
+                out_diag = jnp.zeros_like(diag)
+                
+                sA, sB = max(0, -matrix_offset), max(0, matrix_offset)
+                eA, eB = min(N, N-matrix_offset), min(N, N+matrix_offset)
+
+                out_diag = out_diag.at[sB:eB].add(
+                    diag[sA:eA] * matrix_diag[sB:eB]
+                )
+
+                new_offset = offset + matrix_offset
+
+                vector_list.append(out_diag)
+                offset_list.append(new_offset)
+        
+        out = jnp.vstack(vector_list)
+        
+        return out, offset_list 
+
+        
+    
+
+    def _dag(self):
+        """Returns the hermitian conjugate, call to_dense() to visualize."""
+        self.offsets = tuple(-1 * jnp.array(self.offsets))
+        self.diags = jnp.conjugate(self.diags)
+
+    @functools.partial(jax.jit, static_argnums=(0,))
+    def _add(self, matrix: Array) -> Array:
+        """Sparse + Dense only using information about diagonals & offsets."""
+        N = matrix.shape[0]
+        for offset, diag in zip(self.offsets, self.diags):
+            start = max(0, abs(offset))
+            end = min(N, N + abs(offset))
+            i = jnp.arange(end - start) if offset > 0 else jnp.arange(start, end)
+            j = jnp.arange(start, end) if offset > 0 else jnp.arange(end - start)
+
+            matrix = matrix.at[i, j].add(diag[start:end])
+
+        return matrix
+
+    def __matmul__(self, matrix):
+        if isinstance(matrix, Array):
+            return self._matmul(direction="left", matrix=matrix)
+
+        elif isinstance(matrix, SparseDIA):
+            diags, offsets = self._diamul(matrix=matrix)
+            # offsets = tuple([offset.item() for offset in offsets])
+            # return SparseDIA(diags, offsets)
+
+            diags = jnp.array(diags)
+            offsets = jnp.array(offsets)
+
+            mask = jnp.any(diags != 0, axis=1)
+            diags = diags[mask]
+            offsets = offsets[mask]
+
+            unique_offsets, indices = jnp.unique(offsets, return_inverse=True)
+            result = jnp.zeros((len(unique_offsets), diags.shape[1]))
+
+            for i, offset in enumerate(unique_offsets):
+                result = result.at[i].set(jnp.sum(diags[indices == i], axis=0))
+
+            
+            return SparseDIA(result, tuple([offset.item() for offset in unique_offsets]))
+        
+    def __rmatmul__(self, matrix):
+        if isinstance(matrix, Array):
+            return self._matmul(direction="right", matrix=matrix)
+        
+    def __getitem__(self, index):
```

Comment:
`__getitem__` should support slicing. So we should be able to write these kinds of things:
 - `sparsemat[3:7, 2]`
 - `sparsemat[3:7, 2:4]`
 - `sparsemat[3:7, :]`
 - `sparsemat[..., 2:4]`
 - `sparsemat[None, :, :]`

Maybe the last one is a bit harder though.

Note that all of these might end being quite complicated to implement. Maybe we should remove `__getitem__` for now and implement it properly in a subsequent PR.